File: ts/test/veteval.test.ts
/* Copyright (c) 2025 Richard Rodger, MIT License */


import { describe, test } from 'node:test'
import * as Assert from 'node:assert'
import * as Fs from 'node:fs'
import * as Path from 'node:path'

import { Aontu } from '../dist/aontu'
import { vet } from '../dist/vet'


const SPEC_DIR = Path.join(__dirname, '..', '..', 'test', 'spec')


function unescape(s: string): string {
  let out = ''
  for (let i = 0; i < s.length; i++) {
    const c = s[i]
    if ('\\' === c && i + 1 < s.length) {
      const n = s[++i]
      out += 'n' === n ? '\n' : 't' === n ? '\t' : n
    }
    else {
      out += c
    }
  }
  return out
}


type VetRow = {
  file: string, name: string, schema: string, data: string, opts: any
}


function loadVetRows(): VetRow[] {
  const rows: VetRow[] = []
  for (const file of Fs.readdirSync(SPEC_DIR).filter(
    (f) => f.endsWith('.tsv')).sort()) {
    const text = Fs.readFileSync(Path.join(SPEC_DIR, file), 'utf8')
    for (const line of text.split('\n').map((l) => l.replace(/\r$/, ''))) {
      if ('' === line || line.startsWith('#')) {
        continue
      }
      const parts = line.split('\t')
      if ('vet' !== parts[1] || parts.length < 5) {
        continue
      }
      const expect = JSON.parse(unescape(parts[4]))
      const opts = expect.opts ?? {}
      if (null != opts.at || true === opts.closed ||
        true === opts.partial || null != opts.maxErrors) {
        continue
      }
      // A row whose source names the shared fixtures loads files; the
      // one-document form would have to resolve them from a different
      // base, which is a difference in the TEST rather than in the
      // engines.
      const schema = unescape(parts[2])
      const data = unescape(parts[3])
      if (schema.includes('__FIXTURES__') || data.includes('__FIXTURES__')) {
        continue
      }
      rows.push({ file, name: parts[0], schema, data, opts })
    }
  }
  return rows
}


// Does the one document stand up: does it evaluate to a concrete
// value? `collect` so a failure is recorded rather than thrown, which
// is the same mode vet's own passes use.
function evalAccepts(src: string): boolean {
  const aontu = new Aontu()
  const ctx: any = aontu.ctx({ collect: true })
  let out: any
  try {
    out = aontu.generate(src, undefined, ctx)
  }
  catch {
    return false
  }
  return 0 === ctx.err.length && undefined !== out
}


function union(schema: string, data: string): string | undefined {
  if (statementForm(schema) && statementForm(data)) {
    return schema + '\n' + data + '\n'
  }
  if (schema.includes('$.') || data.includes('$.')) {
    return undefined
  }
  return wrap(schema) + '\n' + wrap(data) + '\n'
}


// Written as key statements at the root, rather than as one literal.
function statementForm(src: string): boolean {
  const t = src.trim()
  if (t.startsWith('{') || t.startsWith('[')) {
    return false
  }
  const aontu = new Aontu()
  const ctx: any = aontu.ctx({ collect: true })
  try {
    return true === (aontu.unify(src, undefined, ctx) as any)?.isMap
  }
  catch {
    return false
  }
}


function wrap(src: string): string {
  const t = src.trim()
  if (t.startsWith('{') || t.startsWith('[')) {
    return 'veteval: ' + t
  }
  return statementForm(src)
    ? 'veteval: {\n' + src + '\n}'
    : 'veteval: (' + t + ')'
}


describe('vet-equals-eval', () => {

  const rows = loadVetRows()

  test('the-corpus-is-not-empty', () => {
    // A filter that quietly matched nothing would make every assertion
    // below vacuous, and a vacuous differential check is worse than
    // none: it reads as coverage.
    Assert.ok(20 < rows.length, 'vet rows found: ' + rows.length)
  })

  test('vet-and-eval-agree-on-accept-reject', () => {
    const disagree: string[] = []
    let skipped = 0

    for (const row of rows) {
      const report = vet(row.schema, row.data,
        { ...row.opts, schemaUrl: 'schema', dataUrl: 'data' } as any)
      const vetAccepts = 'valid' === report.verdict

      const one = union(row.schema, row.data)
      if (null == one) {
        skipped++
        continue
      }
      const evalOk = evalAccepts(one)

      if (vetAccepts !== evalOk) {
        disagree.push(
          `${row.file}:${row.name}` +
          ` vet=${report.verdict}` +
          ` eval=${evalOk ? 'generates' : 'refuses'}` +
          ` | schema: ${JSON.stringify(row.schema)}` +
          ` | data: ${JSON.stringify(row.data)}`)
      }
    }

    Assert.deepEqual(disagree, [],
      'vet and eval disagree on ' + disagree.length + ' row(s):\n' +
      disagree.join('\n'))

    // A skip list that quietly grew to swallow the corpus would leave
    // this green over nothing, so the proportion is bounded too.
    Assert.ok(skipped * 4 < rows.length,
      'too many rows have no single-document spelling: ' +
      skipped + ' of ' + rows.length)
  })

})

File: go/veteval_test.go
/* Copyright (c) 2025 Richard Rodger, MIT License */

package aontu


import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

type vetEvalRow struct {
	file, name, schema, data string
}

func loadVetEvalRows(t *testing.T) []vetEvalRow {
	t.Helper()
	specDir := filepath.Join("..", "test", "spec")
	entries, err := os.ReadDir(specDir)
	if err != nil {
		t.Fatalf("cannot read spec dir %s: %v", specDir, err)
	}
	rows := []vetEvalRow{}
	for _, e := range entries {
		if e.IsDir() || !strings.HasSuffix(e.Name(), ".tsv") {
			continue
		}
		raw, err := os.ReadFile(filepath.Join(specDir, e.Name()))
		if err != nil { //coverage:ignore ReadDir just listed the file
			t.Fatalf("cannot read %s: %v", e.Name(), err)
		}
		for _, line := range strings.Split(string(raw), "\n") {
			line = strings.TrimSuffix(line, "\r")
			if "" == line || strings.HasPrefix(line, "#") {
				continue
			}
			parts := strings.Split(line, "\t")
			if 5 > len(parts) || "vet" != parts[1] {
				continue
			}
			var expect struct {
				Opts struct {
					At        string `json:"at"`
					Closed    bool   `json:"closed"`
					Partial   bool   `json:"partial"`
					MaxErrors *int   `json:"maxErrors"`
				} `json:"opts"`
			}
			if err := json.Unmarshal(
				[]byte(unescapeSpec(parts[4])), &expect); err != nil {
				//coverage:ignore the spec runner rejects a malformed cell first
				t.Fatalf("%s:%s: %v", e.Name(), parts[0], err)
			}
			o := expect.Opts
			if "" != o.At || o.Closed || o.Partial || nil != o.MaxErrors {
				continue
			}
			schema, data := unescapeSpec(parts[2]), unescapeSpec(parts[3])
			// A row naming the shared fixtures loads files, and the
			// one-document form would resolve them from a different
			// base -- a difference in the TEST rather than the engines.
			if strings.Contains(schema, "__FIXTURES__") ||
				strings.Contains(data, "__FIXTURES__") {
				continue
			}
			rows = append(rows,
				vetEvalRow{file: e.Name(), name: parts[0], schema: schema, data: data})
		}
	}
	return rows
}

func vetEvalUnion(schema, data string) string {
	if vetEvalStatements(schema) && vetEvalStatements(data) {
		return schema + "\n" + data + "\n"
	}
	if strings.Contains(schema, "$.") || strings.Contains(data, "$.") {
		return ""
	}
	return vetEvalWrap(schema) + "\n" + vetEvalWrap(data) + "\n"
}

// vetEvalStatements reports whether the source is written as key
// statements at the root, rather than as one literal.
func vetEvalStatements(src string) bool {
	t := strings.TrimSpace(src)
	if strings.HasPrefix(t, "{") || strings.HasPrefix(t, "[") {
		return false
	}
	v, err := New().Unify(src)
	if nil != err || nil == v {
		return false
	}
	_, isMap := v.(*MapVal)
	return isMap
}

func vetEvalWrap(src string) string {
	t := strings.TrimSpace(src)
	if strings.HasPrefix(t, "{") || strings.HasPrefix(t, "[") {
		return "veteval: " + t
	}
	if vetEvalStatements(src) {
		return "veteval: {\n" + src + "\n}"
	}
	return "veteval: (" + t + ")"
}

func TestVetEqualsEval(t *testing.T) {
	rows := loadVetEvalRows(t)

	// A filter that quietly matched nothing would make every assertion
	// below vacuous, and a vacuous differential check is worse than
	// none: it reads as coverage.
	if 20 >= len(rows) {
		t.Fatalf("vet rows found: %d", len(rows))
	}

	disagree := []string{}
	skipped := 0
	for _, row := range rows {
		report := Vet(row.schema, row.data,
			&VetOptions{SchemaURL: "schema", DataURL: "data"})
		vetAccepts := VetValid == report.Verdict

		one := vetEvalUnion(row.schema, row.data)
		if "" == one {
			skipped++
			continue
		}
		out, err := New().Generate(one)
		evalOK := nil == err && nil != out

		if vetAccepts != evalOK {
			verb := "refuses"
			if evalOK {
				verb = "generates"
			}
			disagree = append(disagree,
				row.file+":"+row.name+" vet="+report.Verdict+" eval="+verb+
					" | schema: "+strings.ReplaceAll(row.schema, "\n", "\\n")+
					" | data: "+strings.ReplaceAll(row.data, "\n", "\\n"))
		}
	}

	if 0 < len(disagree) {
		t.Fatalf("vet and eval disagree on %d row(s):\n%s",
			len(disagree), strings.Join(disagree, "\n"))
	}

	// A skip list that quietly grew to swallow the corpus would leave
	// this green over nothing, so the proportion is bounded too.
	if skipped*4 >= len(rows) {
		t.Fatalf("too many rows have no single-document spelling: %d of %d",
			skipped, len(rows))
	}
}
